src/csl/quote.rs

```diff
@@ -0,0 +1,171 @@
+use citationberg::{taxonomy::OtherTerm, TermForm};
+
+use super::{taxonomy::EntryLike, Context};
+
+/// A smart quote substitutor with zero lookahead.
+/// TODO: Move to its own crate.
+#[derive(Debug, Clone)]
+pub struct SmartQuoter {
+    /// The amount of quotes that have been opened.
+    depth: u8,
+    /// Each bit indicates whether the quote at this nesting depth is a double.
+    /// Maximum supported depth is thus 32.
+    kinds: u32,
+}
+
+impl SmartQuoter {
+    /// Start quoting.
+    pub fn new() -> Self {
+        Self { depth: 0, kinds: 0 }
+    }
+
+    /// Determine which smart quote to substitute given this quoter's nesting
+    /// state and the character immediately preceding the quote.
+    pub fn quote<'a>(
+        &mut self,
+        before: Option<char>,
+        quotes: &SmartQuotes<'a>,
+        double: bool,
+    ) -> &'a str {
+        let opened = self.top();
+        let before = before.unwrap_or(' ');
+
+        // If we are after a number and haven't most recently opened a quote of
+        // this kind, produce a prime. Otherwise, we prefer a closing quote.
+        if before.is_numeric() && opened != Some(double) {
+            return if double { "″" } else { "′" };
+        }
+
+        // If we have a single smart quote, didn't recently open a single
+        // quotation, and are after an alphabetic char or an object (e.g. a
+        // math equation), interpret this as an apostrophe.
+        if !double
+            && opened != Some(false)
+            && (before.is_alphabetic() || before == '\u{FFFC}')
+        {
+            return "’";
+        }
+
+        // If the most recently opened quotation is of this kind and the
+        // previous char does not indicate a nested quotation, close it.
+        if opened == Some(double)
+            && !before.is_whitespace()
+            && !is_newline(before)
```

Comment:
It appears that all newline characters in your function are whitespaces.[^1] [^2]

```suggestion
            && !before.is_whitespace()
```

[^1]: https://doc.rust-lang.org/std/primitive.char.html#method.is_whitespace
[^2]: `White_Space` in https://www.unicode.org/Public/UCD/latest/ucd/PropList.txt